### ISSUE TITLE
L013 seems to expect an alias for the `*` when subsequent columns are present

### ISSUE THREAD

**Issue by pwildenhain** [alanmcruickshank/sqlfluff#399]
test.sql

```sql
select
    *,
    'string' as extra_column
from
    my_table
```

```
$ sqlfluff lint test.sql
== [test.sql] FAIL
L:   2 | P:   5 | L013 | Column expression without alias. Use explicit `AS`
                       | clause.
```

The linting passes if I don't include `extra_column` in the `select`

**Comment by alanmcruickshank** [comment]
This is definitely a bug, but I'm curious as to how it happens. I think it must be that there's something fundamental about how the parse tree is structured in the two cases. This one needs more investigation.

Next step here is to run `sqlfluff parse` on each of the two queries and understand why `select * from blah` means that the `*` is not parsed as an expression, but `select *, foo from blah` means that `*` *is* parsed as an expression. 🤔

**Comment by dmateusp** [comment]
I hit the same issue as above, and your comment @alanmcruickshank lead me to try parsing a "working" statement and a statement that fails with the above bug:

```sql
select * from foo
```

```yaml
FileSegment:                                                [0](1, 1, 1)
    StatementSegment:                                       [0](1, 1, 1)
        SelectStatementSegment:                             [0](1, 1, 1)
            SelectClauseSegment:                            [0](1, 1, 1)
                SELECT_KeywordSegment:                      [0](1, 1, 1)        'select'
                [META] Indent:                              [6](1, 1, 7)
                whitespace_RawSegment:                      [6](1, 1, 7)        ' '
                SelectTargetElementSegment:                 [7](1, 1, 8)
                    star_KeywordSegment:                    [7](1, 1, 8)        '*'
                [META] Dedent:                              [8](1, 1, 9)
            whitespace_RawSegment:                          [8](1, 1, 9)        ' '
            FromClauseSegment:                              [9](1, 1, 10)
                FROM_KeywordSegment:                        [9](1, 1, 10)       'from'
                [META] Indent:                              [13](1, 1, 14)
                whitespace_RawSegment:                      [13](1, 1, 14)      ' '
                TableExpressionSegment:                     [14](1, 1, 15)
                    ObjectReferenceSegment:                 [14](1, 1, 15)
                        naked_identifier_ReSegment:         [14](1, 1, 15)      'foo'
                [META] Dedent:                              [17](1, 1, 18)
    newline_RawSegment:                                     [17](1, 1, 18)      '\n'
```

while:

```sql
select *, bar from foo
```

```yaml
FileSegment:                                                [0](1, 1, 1)
    StatementSegment:                                       [0](1, 1, 1)
        SelectStatementSegment:                             [0](1, 1, 1)
            SelectClauseSegment:                            [0](1, 1, 1)
                SELECT_KeywordSegment:                      [0](1, 1, 1)        'select'
                [META] Indent:                              [6](1, 1, 7)
                whitespace_RawSegment:                      [6](1, 1, 7)        ' '
                SelectTargetElementSegment:                 [7](1, 1, 8)
                    star_KeywordSegment:                    [7](1, 1, 8)        '*'
                comma_KeywordSegment:                       [8](1, 1, 9)        ','
                whitespace_RawSegment:                      [9](1, 1, 10)       ' '
                SelectTargetElementSegment:                 [10](1, 1, 11)
                    ObjectReferenceSegment:                 [10](1, 1, 11)
                        naked_identifier_ReSegment:         [10](1, 1, 11)      'bar'
                [META] Dedent:                              [13](1, 1, 14)
            whitespace_RawSegment:                          [13](1, 1, 14)      ' '
            FromClauseSegment:                              [14](1, 1, 15)
                FROM_KeywordSegment:                        [14](1, 1, 15)      'from'
                [META] Indent:                              [18](1, 1, 19)
                whitespace_RawSegment:                      [18](1, 1, 19)      ' '
                TableExpressionSegment:                     [19](1, 1, 20)
                    ObjectReferenceSegment:                 [19](1, 1, 20)
                        naked_identifier_ReSegment:         [19](1, 1, 20)      'foo'
                [META] Dedent:                              [22](1, 1, 23)
    newline_RawSegment:                                     [22](1, 1, 23)      '\n'
```

It doesn't seem like the parser is at fault, (although I'm just discovering the project)

**Comment by alanmcruickshank** [comment]
Thanks @dmateusp ! Yeah that's what I was expecting it to do. So that makes the original query more puzzling.

Even with the original query it's not being parsed as an expression....

```
$ sqlfluff parse test4.sql
=== [test4.sql] ===
FileSegment:                                                [0](1, 1, 1)
    StatementSegment:                                       [0](1, 1, 1)
        SelectStatementSegment:                             [0](1, 1, 1)
            SelectClauseSegment:                            [0](1, 1, 1)
                SELECT_KeywordSegment:                      [0](1, 1, 1)        'select'
                [META] Indent:                              [6](1, 1, 7)
                newline_RawSegment:                         [6](1, 1, 7)        '\n'
                whitespace_RawSegment:                      [7](1, 2, 1)        '    '
                SelectTargetElementSegment:                 [11](1, 2, 5)
                    star_KeywordSegment:                    [11](1, 2, 5)       '*'
                comma_KeywordSegment:                       [12](1, 2, 6)       ','
                newline_RawSegment:                         [13](1, 2, 7)       '\n'
                whitespace_RawSegment:                      [14](1, 3, 1)       '    '
                SelectTargetElementSegment:                 [18](1, 3, 5)
                    quoted_literal_NamedSegment:            [18](1, 3, 5)       "'string'"
                    whitespace_RawSegment:                  [26](1, 3, 13)      ' '
                    AliasExpressionSegment:                 [27](1, 3, 14)
                        AS_KeywordSegment:                  [27](1, 3, 14)      'as'
                        whitespace_RawSegment:              [29](1, 3, 16)      ' '
                        naked_identifier_ReSegment:         [30](1, 3, 17)      'extra_column'
                [META] Dedent:                              [42](1, 3, 29)
            newline_RawSegment:                             [42](1, 3, 29)      '\n'
            FromClauseSegment:                              [43](1, 4, 1)
                FROM_KeywordSegment:                        [43](1, 4, 1)       'from'
                [META] Indent:                              [47](1, 4, 5)
                newline_RawSegment:                         [47](1, 4, 5)       '\n'
                whitespace_RawSegment:                      [48](1, 5, 1)       '    '
                TableExpressionSegment:                     [52](1, 5, 5)
                    ObjectReferenceSegment:                 [52](1, 5, 5)
                        naked_identifier_ReSegment:         [52](1, 5, 5)       'my_table'
                [META] Dedent:                              [60](1, 5, 13)
```

**Comment by alanmcruickshank** [comment]
Hold up. Problem solved. 

At the moment there's no exception for `*` at all in `Rule_L013`. There is however an exception for *scalar* queries e.g. `SELECT 1`. This means at the moment that `SELECT * FROM blah` is ok because it's being treated as a scalar, and that `SELECT *, foo FROM blah` isn't because it's now got more than one column.

Easy solution. `Rule_L013._eval` now just needs to exclude anything ending in `*`. This means `*` *but also* `foo.*`, `foo.bar.*` or `"foo.quoted.*"`.

**Comment by alanmcruickshank** [comment]
@dmateusp , @pwildenhain , either of you want to have a go at this one?

The `_eval` function for `Rule_L013` is so far quite compact and so this would be a fairly good entry point into understanding how that part of the codebase works.

**Comment by dmateusp** [comment]
@alanmcruickshank I'd love to have a look this evening, thank you for the explanation

**Comment by pwildenhain** [comment]
Y'all are too quick on the draw! Would love to contribute in a future issue, thanks for addressing this so quickly 🏋️‍♂️
